File: BOJ/17298. 오큰수/solution.cc
#include <iostream>
#include <vector>
#include <stack>

using namespace std;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);

    int n;
    cin >> n;

    stack<int> s;
    vector<int> vec(n), res(n);

    for (int i = 0; i < n; i += 1) {
        cin >> vec[i];
    }

    for (int i = n - 1; i >= 0; i -= 1) {
        while (!s.empty() && s.top() <= vec[i]) s.pop();

        s.empty() ? res[i] = -1 : res[i] = s.top();
        s.push(vec[i]);
    }

    for (int k: res) cout << k << " ";

    return 0;
}

File: BOJ/17298. 오큰수/solution.js
const [len, nums] = require('fs').readFileSync('/dev/stdin').toString().trim().split('\n');
const solution = (n, arr) => {
    const stack = [];
    let res = [];

    for (let i = n; i > 0; i -= 1) {
        const cur = +arr[i - 1], prev = +arr[i];
        if (cur < prev) {
            stack.push(prev);
            res.push(prev);
        }
        else {
            while (stack.length) {
                let top = stack[stack.length - 1];
                if (top > cur) {
                    res.push(top);
                    break;
                }
                else stack.pop();
            }
            if (!stack.length) res.push(-1);
        }
    }
    return res.reverse().join(' ');
}

console.log(solution(+len, nums.split(' ')));